src/internal_modules/roc_sndio/target_sndfile/roc_sndio/sndfile_sink.cpp

```diff
@@ -0,0 +1,360 @@
+}
+
+core::nanoseconds_t SndfileSink::latency() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: latency(): non-open output file or device");
+    }
+
+    return 0;
+}
+
+bool SndfileSink::has_latency() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: has_latency(): non-open output file or device");
+    }
+
+    return false;
+}
+
+bool SndfileSink::has_clock() const {
+    roc_panic_if(!valid_);
+
+    if (!sndfile_output_) {
+        roc_panic("sndfile sink: has_clock(): non-open output file or device");
+    }
+
+    return !is_file_;
+}
+
+void SndfileSink::write(audio::Frame& frame) {
+    roc_panic_if(!valid_);
+    const audio::sample_t* frame_data = frame.samples();
+    size_t frame_size = frame.num_samples();
+    audio::sample_t* buffer_data = buffer_.data();
+    size_t buffer_pos = 0;
+
+    while (frame_size > 0) {
+        for (; buffer_pos < buffer_size_ && frame_size > 0; buffer_pos++) {
+            buffer_data[buffer_pos] = *frame_data;
+            frame_data++;
+            frame_size--;
+        }
+
+        if (buffer_pos > 0) {
+            if (sf_write_float(sndfile_output_, buffer_data, (sf_count_t)buffer_size_)
+                    == 0
+                && frame_size > 0) {
+                roc_panic(
+                    "sndfile sink: Unable to write entire frame to file. Reached eof");
+            }
+        }
+        buffer_pos = 0;
+    }
+}
+
+bool SndfileSink::setup_buffer_() {
+    buffer_size_ = sample_spec_.ns_2_samples_overall(frame_length_);
+    if (buffer_size_ == 0) {
+        roc_log(LogError, "sndfile sink: buffer size is zero");
+        return false;
+    }
+    if (!buffer_.resize(buffer_size_)) {
+        roc_log(LogError, "sndfile sink: can't allocate sample buffer");
+        return false;
+    }
+
+    return true;
+}
+
+bool SndfileSink::open_(const char* driver, const char* path) {
+    unsigned long in_rate = (unsigned long)sf_info_out_.samplerate;
+
+    if (sf_info_out_.samplerate == 0) {
+        sf_info_out_.samplerate = 48000;
+    }
+
+    unsigned long out_rate = (unsigned long)sf_info_out_.samplerate;
+
+    if (!driver) {
+        if (!detect_file_extension(&driver, path)) {
+            roc_log(LogDebug, "sndfile sink: Driver extension could not be detected");
+            return false;
+        }
+    }
+
+    int bits = 0;
+    if (!map_to_sndfile(sf_info_out_, driver, bits)) {
+        roc_log(LogDebug,
+                "sndfile sink: Cannot find valid subtype format for major format type");
+        return false;
+    }
+
+    sndfile_output_ = sf_open(path, SFM_WRITE, &sf_info_out_);
+    if (!sndfile_output_) {
+        roc_log(LogDebug, "sndfile sink: can't open: driver=%s path=%s", driver, path);
+        return false;
+    }
+
+    sf_command(sndfile_output_, SFC_SET_UPDATE_HEADER_AUTO, NULL, SF_TRUE);
+
+    if (in_rate != 0 && in_rate != out_rate) {
```

Comment:
I've erased the check for if (in_rate != 0 && in_rate != out_rate). This is never a problem. sndfile in general is capable of changing the sample rate of a file, but it would not be happening here, nor would it cause a problem. Thanks.

Reply:
Got it.